pkg/controller/networkpolicy/store/appliedtogroup.go

```diff
@@ -174,6 +174,22 @@ func NewAppliedToGroupStore() storage.Interface {
 			}
 			return []string{atg.Selector.Namespace}, nil
 		},
+		PodIndex: func(obj interface{}) ([]string, error) {
+			atg, ok := obj.(*types.AppliedToGroup)
+			if !ok {
+				return []string{}, nil
+			}
+			keys := make([]string, 0)
+			for _, podSet := range atg.PodsByNode {
+				for _, pod := range podSet {
+					if pod != nil && pod.Pod != nil {
+						name, namespace := pod.Pod.Name, pod.Pod.Namespace
+						keys = append(keys, name+"/"+namespace)
+					}
+				}
+			}
+			return keys, nil
+		},
```

Comment:
@jianjuns @antoninbas I have the feeling that the PodIndex could cause performance regression as it needs to index all Pods once, for a large Group it might be 100k Pods, could we check the result of TestInitXLargeScaleWithSmallNamespaces with this patch?

Reply:
If by patch you mean the antctl query feature, the performance tradeoffs are 

Without indexing

init
100k policies and 100k pods with 1 applied policy per pod takes 12 seconds (1555M max heap)
10k policies and 10k pods with each policy applied to all pods takes 7 seconds (1138M max heap)


With indexing

query
100k policies and 100k pods with 1 applied policy per pod takes .1 seconds for 1000 queries (2793M max heap)
10k policies and 10k pods with each policy applied to all pods takes 13 seconds for 1000 queries (1092M max heap)

init
100k policies and 100k pods with 1 applied policy per pod takes 13 seconds (1849M max heap)
10k policies and 10k pods with each policy applied to all pods takes 11 seconds (1185M max heap)

Reply:
@jakesokol1 are these numbers for `TestInitXLargeScaleWithSmallNamespaces`? If yes, it looks like the impact on the initial computation time is minimal. I will double-check myself.

Reply:
Yes, though my computer is rather slow

Reply:
@tnqn I ran `TestInitXLargeScaleWithSmallNamespaces` on my laptop with / without the news indexes, and this is what I observed (representative results):
```
** WITHOUT NEW INDEX **

--- PASS: TestInitXLargeScaleWithSmallNamespaces (7.15s)
    networkpolicy_controller_perf_test.go:251: Execution metrics of addNamespace, executions: 25000, duration: 971.817374ms
    networkpolicy_controller_perf_test.go:251: Execution metrics of addNetworkPolicy, executions: 75000, duration: 3.421248811s
    networkpolicy_controller_perf_test.go:251: Execution metrics of processNextAppliedToGroupWorkItem, executions: 78037, duration: 3.421270393s
    networkpolicy_controller_perf_test.go:251: Execution metrics of processNextInternalNetworkPolicyWorkItem, executions: 125907, duration: 3.42128457s
    networkpolicy_controller_perf_test.go:251: Execution metrics of addPod, executions: 100000, duration: 1.656492313s
    networkpolicy_controller_perf_test.go:251: Execution metrics of processNextAddressGroupWorkItem, executions: 51940, duration: 3.421308881s
    networkpolicy_controller_perf_test.go:258: Summary metrics:
        NAMESPACES   PODS    NETWORK-POLICIES    TIME(s)    MEMORY(M)    EXECUTIONS    EVENTS(ag, atg, np)
        25000        100000  75000               3.43       1616         455884        182610 150111 182610
PASS
ok  	github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy	7.507s

** WITH NEW INDEX **

--- PASS: TestInitXLargeScaleWithSmallNamespaces (7.45s)
    networkpolicy_controller_perf_test.go:251: Execution metrics of processNextInternalNetworkPolicyWorkItem, executions: 128459, duration: 3.707525129s
    networkpolicy_controller_perf_test.go:251: Execution metrics of processNextAppliedToGroupWorkItem, executions: 75000, duration: 3.707516042s
    networkpolicy_controller_perf_test.go:251: Execution metrics of processNextAddressGroupWorkItem, executions: 50000, duration: 3.707548309s
    networkpolicy_controller_perf_test.go:251: Execution metrics of addNamespace, executions: 25000, duration: 671.716925ms
    networkpolicy_controller_perf_test.go:251: Execution metrics of addPod, executions: 100000, duration: 1.54879582s
    networkpolicy_controller_perf_test.go:251: Execution metrics of addNetworkPolicy, executions: 75000, duration: 3.707490986s
    networkpolicy_controller_perf_test.go:258: Summary metrics:
        NAMESPACES   PODS    NETWORK-POLICIES    TIME(s)    MEMORY(M)    EXECUTIONS    EVENTS(ag, atg, np)
        25000        100000  75000               3.71       1648         453459        182075 150001 182075
PASS
ok  	github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy	7.808s
```

I also wrote a new test, with 500K Pods and 2 NetworkPolicies, each one with a different AppliedToGroup (250K Pods) and a different AddressGroup (250K Pods). So the groups should be 10x the size of the groups for `TestInitXLargeScaleWithSmallNamespaces`. I first waited for all the Pod ADD events to be processed, then created the 2 NetworkPolicies and measured time / memory usage. The memory usage doesn't seem to vary much. Between 2 runs, I can see some large variations, but I see them in both cases (with / without Pod indexer); I think GC has an impact there. Here are representative results:
```
** WITHOUT NEW INDEX **

abas-a01:antrea abas$ go test ./pkg/controller/networkpolicy/... -run=TestInitXLargeScaleWithLargeAddressGroups -v -count=1
=== RUN   TestInitXLargeScaleWithLargeAddressGroups
I0805 13:33:13.445334   45003 networkpolicy_controller.go:955] Starting NetworkPolicy controller
I0805 13:33:13.445502   45003 networkpolicy_controller.go:958] Waiting for caches to sync for NetworkPolicy controller
I0805 13:33:13.445511   45003 networkpolicy_controller.go:970] Caches are synced for NetworkPolicy controller
I0805 13:33:29.602238   45003 networkpolicy_controller.go:627] Creating new internal NetworkPolicy testNamespace/testNP-even-odd
I0805 13:33:29.602281   45003 networkpolicy_controller.go:627] Creating new internal NetworkPolicy testNamespace/testNP-odd-even
I0805 13:33:35.402304   45003 networkpolicy_controller.go:978] Shutting down NetworkPolicy controller
--- PASS: TestInitXLargeScaleWithLargeAddressGroups (24.42s)
    networkpolicy_controller_perf_test.go:369: Execution metrics of addNetworkPolicy, executions: 2, duration: 497.634µs
    networkpolicy_controller_perf_test.go:369: Execution metrics of processNextInternalNetworkPolicyWorkItem, executions: 4, duration: 1.451477612s
    networkpolicy_controller_perf_test.go:369: Execution metrics of processNextAppliedToGroupWorkItem, executions: 2, duration: 1.450528241s
    networkpolicy_controller_perf_test.go:369: Execution metrics of processNextAddressGroupWorkItem, executions: 2, duration: 2.796164525s
    networkpolicy_controller_perf_test.go:376: Summary metrics:
        NAMESPACES   PODS    NETWORK-POLICIES    TIME(s)    MEMORY(M)    EXECUTIONS    EVENTS(ag, atg, np)
        1            500000  2                   2.80       4818         10            7 3 7
PASS
ok  	github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy	25.007s

** WITH NEW INDEX **

abas-a01:antrea abas$ go test ./pkg/controller/networkpolicy/... -run=TestInitXLargeScaleWithLargeAddressGroups -v -count=1
=== RUN   TestInitXLargeScaleWithLargeAddressGroups
I0805 13:31:22.007341   44922 networkpolicy_controller.go:955] Starting NetworkPolicy controller
I0805 13:31:22.007491   44922 networkpolicy_controller.go:958] Waiting for caches to sync for NetworkPolicy controller
I0805 13:31:22.007498   44922 networkpolicy_controller.go:970] Caches are synced for NetworkPolicy controller
I0805 13:31:37.996181   44922 networkpolicy_controller.go:627] Creating new internal NetworkPolicy testNamespace/testNP-even-odd
I0805 13:31:37.996230   44922 networkpolicy_controller.go:627] Creating new internal NetworkPolicy testNamespace/testNP-odd-even
I0805 13:31:45.188865   44922 networkpolicy_controller.go:978] Shutting down NetworkPolicy controller
--- PASS: TestInitXLargeScaleWithLargeAddressGroups (25.11s)
    networkpolicy_controller_perf_test.go:369: Execution metrics of addNetworkPolicy, executions: 2, duration: 457.701µs
    networkpolicy_controller_perf_test.go:369: Execution metrics of processNextInternalNetworkPolicyWorkItem, executions: 4, duration: 2.052929587s
    networkpolicy_controller_perf_test.go:369: Execution metrics of processNextAppliedToGroupWorkItem, executions: 2, duration: 2.051185927s
    networkpolicy_controller_perf_test.go:369: Execution metrics of processNextAddressGroupWorkItem, executions: 2, duration: 4.189271888s
    networkpolicy_controller_perf_test.go:376: Summary metrics:
        NAMESPACES   PODS    NETWORK-POLICIES    TIME(s)    MEMORY(M)    EXECUTIONS    EVENTS(ag, atg, np)
        1            500000  2                   4.19       4803         10            7 3 7
PASS
ok  	github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy	25.666s
```

I think these results are a good indicator of the impact on CPU time for processing AppliedToGroups and AddressGroups. Overall computation time for these 2 policies goes from 2.8s to 4.2s. But we are talking about very large groups here (250K Pods)... There should also be an impact for a large number of groups with fewer Pods, but it may not be as significant (see `TestInitXLargeScaleWithSmallNamespaces`).

@jianjuns / @tnqn I don't know if you think a feature gate is warranted based on these results? The difference in memory usage seems to be small, and the difference in computation time only becomes significant for very large groups?

Reply:
I think the worst case scenario would be to have each Pod be in many different address groups / applied to groups. It may be possible to write a test case for which each Pod is in each AddressGroup and in each AppliedToGroup, and each NetworkPolicy has a different AddressGroup instance and a different AppliedToGroup instance associated with it. Then the memory required for the indexers would be `O(|policies|*|pods|)` I think. But this is the same bound as memory usage without the indexers I believe. This nightmare scenario seems pretty unrealistic as well.

Reply:
@antoninbas @jianjuns I used go benchmark to measure the impact of the index on each storage operation, the result as below:
```
BenchmarkCreateAndUpdateLargeAddressGroupWithPodIndex-4                3         393108249 ns/op        79127104 B/op     367649 allocs/op
BenchmarkCreateAndUpdateLargeAddressGroupWithoutPodIndex-4            12         125893721 ns/op        34861568 B/op        115 allocs/op
BenchmarkCreateLargeAddressGroupWithPodIndex-4                        16         104866260 ns/op        35915733 B/op     152031 allocs/op
BenchmarkCreateLargeAddressGroupWithoutPodIndex-4                    100          26927390 ns/op        16855188 B/op         54 allocs/op
BenchmarkCreateAndUpdateSmallAddressGroupWithPodIndex-4             1056           1747328 ns/op          497166 B/op       3773 allocs/op
BenchmarkCreateAndUpdateSmallAddressGroupWithoutPodIndex-4          4465            418761 ns/op          143911 B/op         69 allocs/op
BenchmarkCreateSmallAddressGroupWithPodIndex-4                      2846            764351 ns/op          251498 B/op       1568 allocs/op
BenchmarkCreateSmallAddressGroupWithoutPodIndex-4                  10000            102219 ns/op           69797 B/op         36 allocs/op
```
Test code is https://github.com/tnqn/antrea/commit/76a1481c5264a524f5134cf62d58d2a413fd7be7.

Reply:
I guess in real environment and TestInitXLargeScaleWithLargeAddressGroups tests, there are other factors that could affect the load of the process and the execution time, like even with pod index it's slower to store a group, it might process groups less times because some tasks were merged, this can be told from https://github.com/vmware-tanzu/antrea/pull/1021/files#r466003019 too:
Without index, processNextAppliedToGroupWorkItem executions: 78037, processNextAddressGroupWorkItem, executions: 51940
With index, processNextAppliedToGroupWorkItem executions: 75000, processNextAddressGroupWorkItem, executions: 50000

I'm not very sure what impact it would have in real clusters, but I'm thinking maybe we could leverage the pod index more than just for policy querier. For example, when it receives a pod event, currently it will iterate all groups or filter groups by namespaces to see how many groups are affected, with pod index, it could be optimized to an index query? I think there's more places can be optimized too.
So my suggestion is we don't add the index given the tight schedule if we want the feature to include it in 0.9.0, and with an optimization topic, we add the index and leverage it for policy computation and do a full performance test to check the impact, what do you think?

Reply:
Thanks @tnqn. The results make sense to me after looking at the implementation of `updateIndices`: https://github.com/kubernetes/client-go/blob/00dbcca6ee44c678754d3f5fda1bd0e704b26fe2/tools/cache/thread_safe_store.go#L256. In particular with the implementation of `deleteFromIndices`, even an update where only one Pod is added or removed from a large address group would be costly, as we would pretty much delete the existing index and rebuild it from scratch, creating a new `Set` object for each Pod...

Your plan works for me but just to be sure and so that @jakesokol1 knows what to do, you are suggesting that for 0.9, we remove the new index altogether and instead use a "naive" implementation where we iterate over all groups in case of a query to determine which ones include the endpoint / Pod?

Reply:
@antoninbas Yes, I think the query wouldn't be really slow even without the pod index in a cluster that has thousands of pods and network policies. And the overhead of the query would be similar to some networkpolicy/addressgroup/appliedtogroup API queries with some extra calculation.

Reply:
Jake measured the difference and IIRC queries without indexing were probably 100x slower without indexing when looking at a large number of Pods / NetworkPolicies. But I agree that it doesn't really matter at this stage and a typical antctl user won't see the difference.

@jakesokol1 could you revert the implementation and remove the indexes for now?